### PR TITLE
[Group] Added group specific functions for message permitted

### LINKED ISSUE #12101: Removed MessageDispatch bypass in ExchangeContext
#### Problem
In `ExchangeContext.cpp` inside the `HandleMessage` function a bypass was added to allow Group Messaging

>    if (!IsGroupExchangeContext())
    {
        ReturnErrorOnFailure(
            mDispatch->OnMessageReceived(messageCounter, payloadHeader, peerAddress, msgFlags, GetReliableMessageContext()));
    }

As @bzbarsky-apple said on pr #12052 

> This is bypassing the MessagePermitted checks that mDispatch does. Why is that OK?
>
> It seems like we should either have a different type of dispatch for group exchanges (plausible, actually, since it would no-op OnMessageReceived except for the MessagePermitted checks and complain if you SendMessage) or somehow make our existing dispatches check whether we're doing groupcast.

#### Proposed Solution
Again in the words of @bzbarsky-apple 

> I think the separate dispatch makes the most sense. So create a class for that and set mDispatch to it in ExchangeContext::ExchangeContext if we're a group exchange. I really hope exchanges don't dynamically mutate whether they're group or not....

### PR DESCRIPTION
**Pull request by mkardous-silabs**
#### Problem
* Fixes  #12101 

#### Change overview
* Added group specific functions for onMessageReceived in the ExchangeContext

#### Testing
* Automated testing to validate no regressions were caused

### CODE REVIEW

**Review by jepenven-silabs — APPROVED** [review]

File: src/messaging/ExchangeMessageDispatch.cpp [hunk 1]
```diff
@@ -137,5 +138,37 @@ CHIP_ERROR ExchangeMessageDispatch::OnMessageReceived(uint32_t messageCounter, c
...
+bool ExchangeMessageDispatch::GroupMessagePermitted(uint16_t protocol, uint8_t type)
+{
+    switch (protocol)
+    {
+    case Protocols::InteractionModel::Id.GetProtocolId():
+        switch (type)
+        {
+        case static_cast<uint8_t>(Protocols::InteractionModel::MsgType::WriteRequest):
+        case static_cast<uint8_t>(Protocols::InteractionModel::MsgType::InvokeCommandRequest):
+            return true;
```
Comment (tcarmelveilleux) [inline src/messaging/ExchangeMessageDispatch.cpp:160]: This is layer crossing. This should be done at the IM layer in OnMessageReceived. Please remove this as it couples MessageLayer *directly* with IM, which is neither necessary nor allowed.

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
- [x] Let IM layer determine which types of group messages are permitted

File: src/messaging/ExchangeMessageDispatch.cpp (hunk 1 above)
Comment (mkardous-silabs) [reply src/messaging/ExchangeMessageDispatch.cpp:160]: I agree with the comment, but the goal of this PR was to address Boris's comments on having the flow go through `MessagePermited`. By moving these changes to the IM layer, this PR and the issue both become pointless. 
@bzbarsky-apple are you ok with keeping the code as is in the `ExchangeContext `?

A task already exist in order to add a check in IM.  #11743

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
Fast tracking, given this has had enough time for review, also looks like comments will be addressed in a followup issue.

**Comment by woody-apple** [comment]
> [x] Let IM layer determine which types of group messages are permitted
OK to resolve in filed issue.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Marking "changes requested" for the moment so it doesn't get merged before I get a chance to review tonight.

File: src/messaging/ExchangeContext.cpp [hunk 2]
```diff
@@ -447,7 +447,12 @@ CHIP_ERROR ExchangeContext::HandleMessage(uint32_t messageCounter, const Payload
     });
 
     // TODO : Remove this bypass for group as to perform the MessagePermitted function Issue # 12101
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeContext.cpp:449]: This TODO comment should go away?  And the todo issue be linked to this PR?

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request closed without merging**